### DISCUSSION TITLE
how to send a char array with dbcppp

### DISCUSSION THREAD

**Discussion by janaranjanahl** [xR3b0rn/dbcppp#123] [Q&A]
Hi,
I am trying to send char values over CAN. I could not get it working when over 6 bytes. what is the proper way to do it using dbcppp.

```
    char temp_id[8] = "sam123";
    char dbl_to_str[8];
    uint64_t id = 0;
    memcpy(&evccid, temp_evcc_id, 8);
    dbcppp_SignalPhysToRaw(id_sig, id );
    memcpy(dbl_to_str, &id, 8);
    printf("Double to String %s\n", dbl_to_str);
```
this worked

when temp_id is 7 bytes it does not work and show a offset in the double value as below.
`  char temp_id[8] = "sam1234";`

what I am doing wrong here? 

what is the best way to send a mac address or a string over CAN with DBCPPP?

**Comment by xR3b0rn** [comment]
It depends on your DBC... And to copy raw values into your buffer, you should use the `dbcppp_SignalEncode`-function.

**Reply by janaranjanahl** [reply]
Thanks for fast reply and suggestion. here is the particular entry in dbc
```
BO_ 2566059006 PGN_ID: 8 EVCC
 SG_ EVCC_ID : 0|64@1+ (1,0) [0|255] "" EV
```

**Comment by xR3b0rn** [comment]
```
char buffer[8];
dbcppp_SignalEncode(id_sig, id);
```
(You dont have to invoke `SignalPhysToRaw`, since your factor is 1 and offset is 0)

**Reply by janaranjanahl** [reply]
This will work.

But for me it does not work as I can not call `dbcppp_SignalEncode `directly.

in my setup, I fill data array of double type with CAN data and when the proper time interval is met send out data with 
```
                       struct can_frame to_send;
			uint64_t raw = dbcppp_SignalPhysToRaw(array[index].spns[i].sig, array[index].spns[i].phys);
			dbcppp_SignalEncode(array[index].spns[i].sig, raw, to_send);
```

any suggestion that I can copy char buffer[8] to double value? and go though following step.

`uint64_t raw = dbcppp_SignalPhysToRaw(array[index].spns[i].sig, array[index].spns[i].phys);
`

**Reply by xR3b0rn** [reply]
You can write your own `SignalPhysToRaw` function.

**Reply by MichaelBMiner** [reply]
Our code is simple:
```
char expected_str[8] = "smg1234";
	uint64_t phys = 0;
	uint64_t temp = 0;
	memcpy(&phys, expected_str, 8);

	for(int i = 0 ; i<strlen(expected_str) ; i++) {
		printf("%c %d\n", expected_str[i], (int)(expected_str[i]));
		temp |= (int)(expected_str[i]);
		temp = temp << 8;
	}

	printf("%" PRIu64 "\n", phys);
	printf("%" PRIu64 "\n", temp);


	can_hal_set_spn_evcc_id(temp);
```

We want to send 8 bytes of a string over CAN. This temp value is a uint64_t it is turned into a double with the `can_hal_set_spn_evcc_id(temp);` function call. That temp value is stored in a signals phys data holder and sent out using

```
uint64_t raw = dbcppp_SignalPhysToRaw(dbc_id_info_array[index].spns[i].sig, dbc_id_info_array[index].spns[i].phys);
			dbcppp_SignalEncode(dbc_id_info_array[index].spns[i].sig, raw, to_send.data);
```

If we change the type of temp to double then the dbcppp_SignalPhysToRaw just returns a 0. Is there an appropriate way to do this without writing our own implementation?

**Reply by MichaelBMiner** [reply]
It seems this may be a decode issue with cantools. If I send the data char expected_str[8] = "12345678"; as 3689632501694201856, on the other end I receive Timestamp: 1666716604.949046    ID: 18f2f3fe    X                DLC:  8    00 00 38 37 36 35 34 33     Channel: can0 bytearray(b'\x00\x00876543') {'EVCC_ID': 3689632501694201856}.

The RX and TX data are the same but the RX byte array is incorrect.

**Reply by xR3b0rn** [reply]
`dbcppp` assumes every physical value to be a `double`, there is no way to change that.
